apis/kueue/v1beta1/clusterqueue_types.go

```diff
@@ -0,0 +1,228 @@
+	// If null, it means that there is no borrowing limit.
+	// If not null, it must be non-negative.
+	// borrowingLimit must be null if spec.cohort is empty.
+	// +optional
+	BorrowingLimit *resource.Quantity `json:"borrowingLimit,omitempty"`
+}
+
+// ResourceFlavorReference is the name of the ResourceFlavor.
+type ResourceFlavorReference string
+
+// ClusterQueueStatus defines the observed state of ClusterQueue
+type ClusterQueueStatus struct {
+	// flavorUsage are the used quotas by flavor currently in use by the workloads
+	// assigned to this ClusterQueue.
+	// +listType=map
+	// +listMapKey=name
+	// +kubebuilder:validation:MaxItems=16
+	// +optional
+	FlavorUsage []FlavorUsage `json:"flavorUsage"`
+
+	// PendingWorkloads is the number of workloads currently waiting to be
+	// admitted to this clusterQueue.
+	// +optional
+	PendingWorkloads int32 `json:"pendingWorkloads"`
+
+	// AdmittedWorkloads is the number of workloads currently admitted to this
+	// clusterQueue and haven't finished yet.
+	// +optional
+	AdmittedWorkloads int32 `json:"admittedWorkloads"`
+
+	// conditions hold the latest available observations of the ClusterQueue
+	// current state.
+	// +optional
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+
+type FlavorUsage struct {
+	// name of the flavor.
+	Name string `json:"name"`
+
+	// resources lists the quota usage for the resources in this flavor.
+	// +listType=map
+	// +listMapKey=name
+	// +kubebuilder:validation:MaxItems=16
+	Resources []ResourceUsage `json:"resources"`
+}
+
+type ResourceUsage struct {
+	// name of the resource
+	Name corev1.ResourceName `json:"name"`
+
+	// Total is the total quantity of the resource used, including resources
+	// borrowed from the cohort.
+	Total resource.Quantity `json:"total,omitempty"`
+
+	// Borrowed is the used quantity past the min quota, borrowed from the cohort.
+	Borrowed resource.Quantity `json:"borrowing,omitempty"`
+}
+
+const (
+	// ClusterQueueActive indicates that the ClusterQueue can admit new workloads and its quota
+	// can be borrowed by other ClusterQueues in the same cohort.
+	ClusterQueueActive string = "Active"
+)
+
+//+kubebuilder:object:root=true
+//+kubebuilder:resource:scope=Cluster,shortName={cq}
+//+kubebuilder:subresource:status
+//+kubebuilder:printcolumn:name="Cohort",JSONPath=".spec.cohort",type=string,description="Cohort that this ClusterQueue belongs to"
+//+kubebuilder:printcolumn:name="Strategy",JSONPath=".spec.queueingStrategy",type=string,description="The queueing strategy used to prioritize workloads",priority=1
+//+kubebuilder:printcolumn:name="Pending Workloads",JSONPath=".status.pendingWorkloads",type=integer,description="Number of pending workloads"
+//+kubebuilder:printcolumn:name="Admitted Workloads",JSONPath=".status.admittedWorkloads",type=integer,description="Number of admitted workloads that haven't finished yet",priority=1
+//+kubebuilder:storageversion
```

Comment:
Do you plan to add the conversion webhooks?

Reply:
Most likely not. But that's part of the justification of going to beta, that from now on we would have a conversion webhook if there are breaking changes.
But if you know of users running in prod, it would be good to have their feedback.

Reply:
> Most likely not. But that's part of the justification of going to beta: that from now on we would have a conversion webhook if there are breaking changes from v1beta1 to v1beta2 or v1.

It sounds good to me.

> But if you know of users running in prod, it would be good to have their feedback.

I haven't known of users running in production, ever.